core/shared/src/main/scala/org/http4s/Uri.scala

```diff
@@ -339,12 +339,14 @@ object Uri extends UriPlatform {
     private def doEquals(path: Path): Boolean =
       this.segments == path.segments && path.absolute == this.absolute && path.endsWithSlash == this.endsWithSlash
 
-    override def hashCode(): Int = {
-      var hash = segments.hashCode()
-      hash += 31 * absolute.hashCode()
-      hash += 31 * endsWithSlash.hashCode()
-      hash
-    }
+    override def hashCode(): Int =
```

Comment:
Since we're breaking how we calculate hashCode, I suggest calculating it in a manner as Scala 2.13 does for case classes. It should bring much fewer collisions. WDYT?

Reply:
I recently did the same for [`HttpVersion` hashCode()](https://github.com/http4s/http4s/pull/6544/files#diff-d2517b0e3afb33d3674f2249a71dd41ec8a966644a9930d334169e3753728c8cR79-R84), I can say it produces an identical result with Scala 2.13.

Reply:
cool! Do you have a preference over whether to implement it exactly as you did for `HttpVersion` or is the snippet I posted above fine?

Reply:
I don't think `System.identityHashCode` is right, it's based on the underlying reference. E.g.
```
val x = Some(1)
val y = Some(1)
```

`x` and `y` would not have the same `identityHashCode` because they are different references.

Reply:
Gotcha. Let me fix that